src/json.c

```diff
@@ -42,76 +43,131 @@ int GetJSONAPIs(RedisModuleCtx *ctx, int subscribeToModuleChange) {
 
-int JSON_GetRedisModuleString(RedisModuleCtx *ctx, RedisJSON json, RedisModuleString **str) {
-  size_t len;
-  const char *cstr;
-  long long intval;
-  double dblval;
-  int boolval;
 
+int FieldSpec_CheckJsonType(FieldType fieldType, JSONType type) {
+  int rv = REDISMODULE_ERR;
+  switch (type) {
+  // TEXT, TAG and GEO fields are represented as string 
+  case JSONType_String:
+    if (fieldType == INDEXFLD_T_FULLTEXT || fieldType == INDEXFLD_T_TAG || fieldType == INDEXFLD_T_GEO) {
+      rv = REDISMODULE_OK;
+    }
+    break;
+  // NUMERIC field is represented as either integer or double 
+  case JSONType_Int:
+  case JSONType_Double:
+    if (fieldType == INDEXFLD_T_NUMERIC) {
+      rv = REDISMODULE_OK;
+    }
+    break;
+  // Boolean values can be represented only as TAG 
+  case JSONType_Bool:    
+    if (fieldType == INDEXFLD_T_TAG) {
+      rv = REDISMODULE_OK;
+    }
+    break;
+  // An object or null type are not supported
+  case JSONType_Array:
+  case JSONType_Object:
+  case JSONType_Null:
+  case JSONType__EOF:
+    break;
+  }
+
+  return rv;
+}
+
+int JSON_StoreInDocField(RedisJSON json, JSONType jsonType, struct DocumentField *df) {
   int rv = REDISMODULE_OK;
 
-  switch (japi->getType(json)) {
+  int boolval;
+  const char *str;
+  long long intval;
+
+  switch (jsonType) {
     case JSONType_String:
-      japi->getString(json, &cstr, &len);
-      *str = RedisModule_CreateString(ctx, cstr, len);
+      japi->getString(json, &str, &df->strlen);
+      df->strval = rm_strndup(str, df->strlen);
+      df->unionType = FLD_VAR_T_CSTR;
       break;
     case JSONType_Int:
       japi->getInt(json, &intval);
-      *str = RedisModule_CreateStringFromLongLong(ctx, intval);
+      df->numval = intval;
+      df->unionType = FLD_VAR_T_NUM;
       break;
     case JSONType_Double:
-      japi->getDouble(json, &dblval);
-      *str = RedisModule_CreateStringFromDouble(ctx, dblval);
+      japi->getDouble(json, &df->numval);
+      df->unionType = FLD_VAR_T_NUM;
       break;
     case JSONType_Bool:
       japi->getBoolean(json, &boolval);
       if (boolval) {
-        *str = RedisModule_CreateString(ctx, "true", strlen("true"));
+        df->strlen = strlen("true");
+        df->strval = rm_strndup("true", df->strlen);
       } else {
-        *str = RedisModule_CreateString(ctx, "false", strlen("false"));
+        df->strlen = strlen("false");
+        df->strval = rm_strndup("false", df->strlen);
       }
+      df->unionType = FLD_VAR_T_CSTR;
       break;
-    case JSONType_Object:
     case JSONType_Array:
+    case JSONType_Object:
     case JSONType_Null:
       rv = REDISMODULE_ERR;
+      break;
     case JSONType__EOF:
       RS_LOG_ASSERT(0, "Should not happen");
   }
+
   return rv;
 }
 
-int FieldSpec_CheckJsonType(FieldType fieldType, JSONType type) {
-  int rv = REDISMODULE_ERR;
-  switch (type) {
+int JSON_GetRedisModuleString(JSONResultsIterator jsonIter, size_t len,
```

Comment:
I think this is no longer a good name to indicate what the function does no?

Reply:
Good point, how about `JSON_LoadDocumentField`?